File: src/main/java/com/axellience/vuegwt/client/jsnative/JsTools.java
package com.axellience.vuegwt.client.jsnative;

import com.axellience.vuegwt.client.jsnative.types.JsIterator;
import jsinterop.annotations.JsMethod;

/**
 * Source: https://github.com/ltearno/angular2-gwt/
 */
public class JsTools
{
    @JsMethod(namespace = "console")
    public static native void log(String message);

    @JsMethod(namespace = "axellience")
    public static native Object getNativeType(String typeName);

    @JsMethod(namespace = "axellience")
    public static native Object getWindow();

    @JsMethod(namespace = "axellience")
    public static native boolean objectHasProperty(Object object, String property);

    @JsMethod(namespace = "axellience")
    public static native Object getObjectProperty(Object object, String property);

    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, Object value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, boolean value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, byte value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, char value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, float value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, int value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, short value);
    @JsMethod(namespace = "axellience")
    public static native void setObjectProperty(Object object, String property, double value);

    @JsMethod(namespace = "axellience")
    public static native void unsetObjectProperty(Object object, String property);

    @JsMethod(namespace = "axellience")
    public static native <T> T getArrayItem(Object array, int index);

    @JsMethod(namespace = "axellience")
    public static native void setArrayItem(Object array, int index, Object value);

    @JsMethod(namespace = "axellience")
    public static native <T> JsIterator<T> getObjectIterator(Object object);

    @JsMethod(namespace = "axellience")
    public static native <T> T convertObject(String prototypeName, Object template);

    @JsMethod(namespace = "window.history")
    public static native void historyGoBack();

    public static <T> T get(Object o, int index)
    {
        return getArrayItem(o, index);
    }

    public static <T> void set(Object o, int index, T value)
    {
        setArrayItem(o, index, value);
    }

    @SuppressWarnings("unchecked")
    public static <T> T get(Object o, String propertyName)
    {
        return (T) getObjectProperty(o, propertyName);
    }

    public static void set(Object o, String propertyName, Object value)
    {
        setObjectProperty(o, propertyName, value);
    }

}

File: src/main/resources/com/axellience/vuegwt/client/resources/js-tools.js
(function (context) {
	if (!context.VueGWT)
		context.VueGWT = {};
	if (!context.VueGWT.jsTools)
		context.VueGWT.jsTools = {};

	const jsTools = context.VueGWT.jsTools;
	/**
	 * Source: https://github.com/ltearno/angular2-gwt/
	 */
	jsTools.defineProperty = function (object, name, definition) {
		const def = {};
		if ('get' in definition) {
			def.get = function () {
				return definition.get(this);
			};
		}

		if ('set' in definition) {
			def.set = function (value) {
				definition.set(this, value);
			}
		}

		Object.defineProperty(object, name, def);
	};
	jsTools.getWindow = function () {
		return context;
	};
	jsTools.objectHasProperty = function (object, property) {
		return property in object;
	};
	jsTools.getObjectProperty = function (object, property) {
		return object[property];
	};
	jsTools.setObjectProperty = function (object, property, value) {
		object[property] = value;
	};
	jsTools.unsetObjectProperty = function (object, property) {
		delete object[property];
	};
	jsTools.setObjectFunction = function (object, property, func) {
		object[property] = function () {
			return func.exec.apply(this, arguments);
		};
	};
	jsTools.getArrayItem = function (array, index) {
		return array[index] || null;
	};
	jsTools.setArrayItem = function (array, index, value) {
		array[index] = value;
	};
	jsTools.call = function (func, thisArg, args) {
		return func.call(thisArg, args);
	};
	jsTools.callOnContext = function (func, args) {
		return func(args);
	};
	jsTools.getDeepValue = function (obj, path) {
		for (let i = 0, pathObj = path.split('.'), len = pathObj.length; i < len; i++) {
			obj = obj[pathObj[i]];
		}
		return obj;
	};
	jsTools.createFunction = function (body) {
		return new Function(body);
	};
	jsTools.getFunctionBody = function (myFunction) {
		// Get content between first { and last }
		const m = myFunction.toString().match(/\{([\s\S]*)\}/m)[1];
		// Strip comments
		return m.replace(/^\s*\/\/.*$/mg, '').trim();
	};
	jsTools.debugger = function () {
		debugger;
	};
	jsTools.delete = function (object, property) {
		return object[property];
	};
	jsTools.isUndefined = function (value) {
		return value === undefined;
	};
})(window);